### ISSUE TITLE
Polynomial Class abstraction

### ISSUE THREAD

**Issue by FelixBenning** [f0uriest/orthax#46]
While orthax implements all the functions from `numpy.polynomials` it does not implement the Polynomial class abstractions. I realize that this is likely because jax encourages a functional approach, but I catch myself reimplementing such abstractions over and over

<details><summary>Example which grew in my code base</summary>

```python
def chebfun(coeffs):
    """Returns the function
        chebfun(rho) = sum_n(coeffs[n] * T_n(rho))

    where T_n are chebychev polynomials.
    Note that chebfun.coeffs returns the original coefficients.
    """

    def fun(cos_angles, diff=0):
        """Compute the covariance kernel for a given cosine angle. Assuming fixed radii.

        Args:
            cos_angles: Array or scalar of cosine angles (cos(<x,y>/|x||y|)).
            diff: Order of derivative to compute. Default is 0 (no derivative).
        """
        return chebyshev.chebval(x=cos_angles, c=chebyshev.chebder(coeffs, m=diff))

    fun.coeffs = coeffs
    return fun
```

This is basically just a class at this point since I am attaching the coefficients to the function for later use.

</details>

In general I feel like polynomials are a (rare?) example where object oriented programming really shines because polynomials are naturally [Function objects](https://en.wikipedia.org/wiki/Function_object). You can add them, evaluate them, you can take derivatives, etc.
So if you are okay with the feature creep I would be happy to start a pull request.

**Comment by FelixBenning** [comment]
I guess a question is how to make such a class play nicely with jax. I would try to read the documentation on
[Pytree](https://docs.jax.dev/en/latest/pytrees.html) objects and try to make sure it does.

**Comment by f0uriest** [comment]
I agree this is a useful abstraction. The main reason I didn't implement it was as you point out, jax is not always particularly friendly with custom classes. There are certainly ways to do it, but it often isnt as idiomatic as in "regular" python. In particular, avoiding cache misses on jit compiled functions and making sure all the usual transformations work correctly can be tricky. ie `jax.grad(Polynomial)` will return another `Polynomial` instance, but its not really a polynomial, its the gradient of the coefficients. Similarly, `jax.vmap(Polynomial)` would give back a single `Polynomial` object with multiple sets of coefficients, and its not clear in that case how evaluation etc should work. 

If you can come up with a solution that seems to cover all the boxes I'd be happy to take a PR, but I suspect that covering every use case is quite difficult so this might be better left as an abstraction that users can built themselves for their particular case.
